File: Astro-TFG/public/scripts/2-procesamiento-datos/prueba-2/version-2/python/main.py
import js  # type: ignore
from pyscript import display, PyWorker
import json
import time
import asyncio

workers = []
workers_ready = False


async def initialize_worker():
    global workers, workers_ready

    try:
        num_workers = max(int(js.document.querySelector(
            "#parallel-workers-pyscript").value) or 1, 1)

        if len(workers) > num_workers:
            workers = workers[:num_workers]

        elif len(workers) < num_workers:
            to_create = num_workers - len(workers)
            origin = js.window.location.origin
            py_path = js.window.document.body.dataset.pyPath

            for _ in range(to_create):
                worker = PyWorker(
                    f"{origin}{py_path}worker.py",
                    type="pyodide",
                    config=f"{origin}{py_path.replace('python', 'json')}pyscript-worker.json"
                )
                await worker.ready
                workers.append(worker)
        workers_ready = True
    except Exception as e:
        display(f"Error creating workers: {e}", target="pyscript-output")

asyncio.create_task(initialize_worker())


async def run_py_benchmark(event):
    global workers
    try:
        num_executions = int(js.document.querySelector(
            "#num-executions-pyscript").value)
        num_workers = int(js.document.querySelector(
            "#parallel-workers-pyscript").value)

        await initialize_worker()

        start_time = time.perf_counter()

        tasks = []
        for i in range(num_executions):
            worker = workers[i % num_workers]
            tasks.append(worker.sync.do_statistical_analysis(100_000))

        results_list = await asyncio.gather(*tasks)
        accumulated = {
            'create': {'time': 0.0, 'memory': 0.0},
            'sum': {'time': 0.0, 'memory': 0.0},
            'mean': {'time': 0.0, 'memory': 0.0},
            'std': {'time': 0.0, 'memory': 0.0},
            'total_per_execution': 0.0,
        }

        for result in results_list:
            data = json.loads(result)
            for op in ['create', 'sum', 'mean', 'std']:
                accumulated[op]['time'] += data[op]['time']
                accumulated[op]['memory'] += data[op]['memory']
            accumulated['total_per_execution'] += data['total']['time']

        results = {op: {
            'time': accumulated[op]['time'] / num_executions,
            'memory': accumulated[op]['memory'] / num_executions
        } for op in ['create', 'sum', 'mean', 'std']}

        results['total'] = {
            'average_per_execution': accumulated['total_per_execution'] / num_executions,
            'total_time': (time.perf_counter() - start_time) * 1000
        }

        display_result(results)

    except Exception as e:
        display(f"Error: {str(e)}", target="pyscript-output")


def display_result(results):
    for operation in ['create', 'sum', 'mean', 'std']:
        display(
            f"{operation.upper()} - Time: {results[operation]['time']:.2f} ms | RAM: {results[operation]['memory']: .2f} MB",
            target="pyscript-output"
        )

    display(
        f"Avg request time: {results['total']['average_per_execution']:.2f} ms",
        target="pyscript-output"
    )
    display(
        f"Total ET: {results['total']['total_time']:.2f} ms",
        target="pyscript-output"
    )
    js.hideExecutionLoader()


def js_run_py_benchmark(event):
    js.clearCell('pyscript-output')
    asyncio.ensure_future(run_py_benchmark(None))


js.window.run_py_benchmark = js_run_py_benchmark

File: Astro-TFG/public/scripts/2-procesamiento-datos/prueba-2/version-2/javascript/main.js
let workers = [];
let workersReady = false;

function initializeWorkers() {
    return new Promise((resolve) => {
        const numWorkers =
            parseInt(
                document.getElementById("parallel-workers-javascript").value
            ) || 1;

        if (workers.length > numWorkers) {
            workers = workers.slice(0, numWorkers);
            workersReady = true;
            resolve();
            return;
        }

        const additionalWorkers = numWorkers - workers.length;
        let initialized = 0;

        for (let i = 0; i < additionalWorkers; i++) {
            const worker = new Worker(window.location.origin + window.document.body.dataset.jsPath + "worker.js");

            worker.onmessage = (event) => {
                const { id, results } = event.data;
                if (pendingPromises.has(id)) {
                    pendingPromises.get(id)(results);
                    pendingPromises.delete(id);
                }
            };

            workers.push(worker);
            initialized++;

            if (initialized === additionalWorkers) {
                workersReady = true;
                resolve();
            }
        }

        if (additionalWorkers === 0) {
            workersReady = true;
            resolve();
        }
    });
}
const pendingPromises = new Map();

async function runJsBenchmark(event) {
    try {
        window.clearCell("javascript-output");
        const numExecutions =
            parseInt(
                document.getElementById("num-executions-javascript").value
            ) || 1;
        const numWorkers =
            parseInt(
                document.getElementById("parallel-workers-javascript").value
            ) || 1;

        await initializeWorkers();

        const simulationStartTime = performance.now();
        const promises = [];

        for (let i = 0; i < numExecutions; i++) {
            const worker = workers[i % numWorkers];
            const id = `${i}-${Date.now()}`;

            promises.push(
                new Promise((resolve) => {
                    pendingPromises.set(id, resolve);
                    worker.postMessage({ id, size: 10_000_000 });
                })
            );
        }

        const resultsArray = await Promise.all(promises);
        let accumulated = {
            create: { time: 0, memory: 0 },
            sum: { time: 0, memory: 0 },
            mean: { time: 0, memory: 0 },
            std: { time: 0, memory: 0 },
            total_per_execution: 0,
        };

        resultsArray.forEach((result) => {
            ["create", "sum", "mean", "std"].forEach((op) => {
                accumulated[op].time += result[op].time;
                accumulated[op].memory += result[op].memory;
            });
            accumulated.total_per_execution += result.total.time;
        });

        const results = {};
        ["create", "sum", "mean", "std"].forEach((op) => {
            results[op] = {
                time: accumulated[op].time / numExecutions,
                memory: accumulated[op].memory / numExecutions,
            };
        });

        results.total = {
            average_per_execution:
                accumulated.total_per_execution / numExecutions,
            total_time: performance.now() - simulationStartTime,
        };

        displayResults(results);
    } catch (error) {
        console.error("Worker error:", error);
    }
}

function displayResults(results) {
    const output = document.getElementById("javascript-output");

    for (const [op, data] of Object.entries(results)) {
        if (op !== 'total') {
            const div = document.createElement("div");
            div.textContent =
                `${op.toUpperCase()} - Time: ${data.time.toFixed(2)} ms | RAM: ${data.memory.toFixed(2)} MB`;
            output.appendChild(div);
        }
    }

    const avgTimeDiv = document.createElement("div");
    avgTimeDiv.textContent =
        `Avg request time: ${results.total.average_per_execution.toFixed(2)} ms`;
    output.appendChild(avgTimeDiv);

    const timeTotalDiv = document.createElement("div");
    timeTotalDiv.textContent =
        `Total ET: ${results.total.total_time.toFixed(2)} ms`;
    output.appendChild(timeTotalDiv);
}


window.runJsBenchmark = async function () {
    window.clearCell("javascript-output");
    window.showExecutionLoader();

    await new Promise(requestAnimationFrame);

    try {
        await runJsBenchmark();
    } finally {
        window.hideExecutionLoader();
    }
};